### DISCUSSION TITLE
Automatic URLSessionTaskMetrics performance tracking?

### DISCUSSION THREAD

**Discussion by WFT** [getsentry/sentry-cocoa#3150] [Q&A] [answered]
Does Sentry support gathering URLSessionTaskTransactionMetrics with its automatic URLSession breadcrumbs/performance tracking? If so, how do I enable it? As far as I can tell the default options don't do it.

It would be helpful to track the performance of our API from the perspective of our end users to know at least:
- “secure connection time” AKA tls time
- fetchStartDate -> responseStartDate AKA “time to first byte”

**Comment by philipphofmann** [comment, marked as answer]
No, we don't support that feature at the moment. The URLSessionTaskMetrics works by using a delegate with something like the following
```Swift
guard let imgUrl = URL(string: "https://sentry-brand.storage.googleapis.com/sentry-logo-black.png") else {
    return
}
let session = URLSession(configuration: URLSessionConfiguration.default)
let dataTask = session.dataTask(with: imgUrl) { (_, _, _) in }
dataTask.delegate = Delegate()
dataTask.resume()

class Delegate: NSObject, URLSessionTaskDelegate {
    func urlSession(_ session: URLSession, task: URLSessionTask, didFinishCollecting metrics: URLSessionTaskMetrics) {

    }
}
```

Our NetworkTracking integration swizzles the `NSURLSessionTask` `setState` and `resume` methods.

https://github.com/getsentry/sentry-cocoa/blob/43860453150ea689c604b2a67d940dc675b70cdd/Sources/Sentry/SentryNetworkTrackingIntegration.m#L57-L62

As we have to use a delegate to retrieve the `URLSessionTaskMetrics`, we can't simply set our delegate as we could overwrite existing delegates. Maybe we could set our own delegate and forward requests to the existing delegate, but that sounds a bit hacky. What are your thoughts, @brustolin?

**Reply by WFT** [reply]
Hey, thanks for responding! That makes sense — it is a little hairy if you’re not already intercepting delegate messages. 

Maybe a good place to start would be a method that lets end users who have a delegate already report this info to Sentry?

Something like `SentrySDK.reportURLSessionTaskMetrics(session, task, metrics)` which does the necessary work of pulling out the spanId/traceId, adding all the child spans and their measurements/data. Then end users of the library could create their own delegates which call that.

**Reply by philipphofmann** [reply]
Yes, we could do what you suggested, but the downside is that you still have to do some manual work, and we always try to give you as much as we can out of the box.

**Reply by philipphofmann** [reply]
@WFT, we created an issue for adding a manual API https://github.com/getsentry/sentry-cocoa/issues/3180. We can't give you an ETA yet.

**Reply by WFT** [reply]
Thanks @philipphofmann ! I've already created my own solution for my codebase, so I'm not too worried about the ETA; I just think it would be a good thing for everyone to have!

**Comment by brustolin** [comment]
> Maybe we could set our own delegate and forward requests to the existing delegate

I don't like this solution either, it is a little bit dangerous.
The manual api would be the easiest solution right now. But not the friendliest for the user.
